src/com/phonebookservice/controller/ContactController.java

```diff
@@ -0,0 +1,150 @@
+    }
+
+    /**
+     * Get instance of contact controller.
+     *
+     * @param database the database.
+     * @return contact controller.
+     */
+    public static ContactController getInstance(
+            final IDataAccessAdapter database) {
+        if (singleton == null) {
+            singleton = new ContactController(database);
+        }
+
+        return singleton;
+    }
+
+    /**
+     * get contact.
+     *
+     * @param contactId the contact id.
+     *
+     * @return contact the contact.
+     */
+    @Override
+    public Contact get(final Long contactId) {
+        ContactController.checkContactId(contactId);
+        final Contact contact = database.get(contactId);
+        ContactController.validateExisitingContact(contact);
+        return contact;
+    }
+
+    /**
+     * create contact.
+     *
+     * @param contact the contact
+     */
+    @Override
+    public void create(final Contact contact) {
+        ContactController.validateContact(contact);
+        database.create(contact);
+    }
+
+    /**
+     * update contact.
+     *
+     * @param contactId the contact id.
+     * @param contact   the updates of contact.
+     *
+     * @return contact the updated contact.
+     */
+    @Override
+    public Contact update(final Long contactId, final Contact contact) {
+        ContactController.checkContactId(contactId);
+        ContactController.validateContact(contact);
+        final Contact existedContact = database.get(contactId);
+        ContactController.validateExisitingContact(existedContact);
+        return database.update(contactId, contact);
+    }
+
+    /**
+     * delete contact.
+     *
+     * @param contactId the contact id.
+     */
+    @Override
+    public void delete(final Long contactId) {
+        ContactController.checkContactId(contactId);
+        final Contact contact = database.get(contactId);
+        ContactController.validateExisitingContact(contact);
+        database.delete(contactId);
+
+    }
+
+    /**
+     * get list of contacts based on filters.
+     *
+     * @param paramValues the hashmap of query params.
+     * @return list of contacts.
+     */
+    @Override
+    public List<Contact> getList(
+            final HashMap<EnumQueryParams, QueryParam> paramValues) {
+        throw new UnsupportedOperationException();
+    }
+
+    private static void validateContact(final Contact contact) {
+        if (contact == null) {
+            throw new BadRequestException(ErrorMessages.ERROR_CONTACT_IS_NULL,
+                    ErrorCode.ERROR_CONTACT_MISSING);
+        }
+
+        if (StringUtility.isNullOrEmptyString(contact.getLastName())) {
+            throw new BadRequestException(ErrorMessages.ERROR_LAST_NAME_IS_NULL,
+                    ErrorCode.ERROR_CONTACT_LAST_NAME_MISSING);
+        }
+    }
+
+    private static void validateExisitingContact(final Contact contact) {
+        if (contact == null) {
+            throw new BadRequestException(
```

Comment:
When the resource doesn't exist, are we returning 400 or 404 ??!!! 

what is the bad that customer did here ? Customer input is right.

BadRequest means that customer did input something bad. 
updating non existing object doesn't mean customer did enter bad input

Reply:
it is a 404 status code